FireRender.Maya.Src/FireRenderToonMaterial.cpp

```diff
@@ -0,0 +1,194 @@
+	CHECK_MSTATUS(addAttribute(attr)); \
+	CHECK_MSTATUS(attributeAffects(attr, Attribute::output));
+
+
+MStatus FireMaya::ToonMaterial::initialize()
+{
+	MFnNumericAttribute nAttr;
+
+	Attribute::output = nAttr.createColor("outColor", "oc");
+	MAKE_OUTPUT(nAttr);
+
+	Attribute::showAdvanced = nAttr.create("showAdvanced", "sa", MFnNumericData::kBoolean, false);
+	MAKE_INPUT(nAttr);
+	nAttr.setConnectable(false);
+
+	Attribute::showMixLevels = nAttr.create("showMixLevels", "sml", MFnNumericData::kBoolean, false);
+	MAKE_INPUT(nAttr);
+	nAttr.setConnectable(false);
+
+	Attribute::color = nAttr.createColor("color", "c");
+	MAKE_INPUT(nAttr);
+	CHECK_MSTATUS(nAttr.setDefault(0.8, 0.8, 0.8));
+
+	Attribute::normal = nAttr.createColor("normal", "n");
+	MAKE_INPUT(nAttr);
+	CHECK_MSTATUS(nAttr.setDefault(1.0f, 1.0f, 1.0f));
+
+	Attribute::roughness = nAttr.create("roughness", "r", MFnNumericData::kFloat, 1.0);
+	MAKE_INPUT(nAttr);
+	nAttr.setMin(0.0);
+	nAttr.setMax(1.0);
+
+	Attribute::highlightColor = nAttr.createColor("highlightColor", "hc");
+	MAKE_INPUT(nAttr);
+	CHECK_MSTATUS(nAttr.setDefault(0.8, 0.8, 0.8));
+
+	Attribute::midColor = nAttr.createColor("midColor", "mc");
+	MAKE_INPUT(nAttr);
+	CHECK_MSTATUS(nAttr.setDefault(0.4, 0.4, 0.4));
+
+	Attribute::shadowColor = nAttr.createColor("shadowColor", "sc");
+	MAKE_INPUT(nAttr);
+	CHECK_MSTATUS(nAttr.setDefault(0.0, 0.0, 0.0));
+
+	// Mid Level
+	Attribute::rampPosition1 = nAttr.create("rampPosition1", "rp1", MFnNumericData::kFloat, 0.5);
+	MAKE_INPUT(nAttr);
+	nAttr.setMin(0.0);
+	nAttr.setMax(1.0);
+
+	// Highlight Level
+	Attribute::rampPosition2 = nAttr.create("rampPosition2", "rp2", MFnNumericData::kFloat, 0.8);
+	MAKE_INPUT(nAttr);
+	nAttr.setMin(0.0);
+	nAttr.setMax(1.0);
+
+	// Mid Level Mix
+	Attribute::rampRange1 = nAttr.create("rampRange1", "rr1", MFnNumericData::kFloat, 0.1);
+	MAKE_INPUT(nAttr);
+	nAttr.setMin(0.0);
+	nAttr.setMax(1.0);
+
+	// Highlight Level Mix
+	Attribute::rampRange2 = nAttr.create("rampRange2", "rr2", MFnNumericData::kFloat, 0.1);
+	MAKE_INPUT(nAttr);
+	nAttr.setMin(0.0);
+	nAttr.setMax(1.0);
+
+	// Adding all attributes to the node type
+	addAttribute(Attribute::output);
+
+	ADD_ATTRIBUTE(Attribute::showAdvanced);
+	ADD_ATTRIBUTE(Attribute::showMixLevels);
+
+	ADD_ATTRIBUTE(Attribute::color);
+	ADD_ATTRIBUTE(Attribute::normal);
+	ADD_ATTRIBUTE(Attribute::roughness);
+
+	ADD_ATTRIBUTE(Attribute::highlightColor);
+	ADD_ATTRIBUTE(Attribute::midColor);
+	ADD_ATTRIBUTE(Attribute::shadowColor);
+
+	ADD_ATTRIBUTE(Attribute::rampPosition1);
+	ADD_ATTRIBUTE(Attribute::rampPosition2);
+
+	ADD_ATTRIBUTE(Attribute::rampRange1);
+	ADD_ATTRIBUTE(Attribute::rampRange2);
+
+	return MStatus::kSuccess;
+}
+
+// creates an instance of the node
+void* FireMaya::ToonMaterial::creator()
+{
+	return new ToonMaterial;
+}
+
+MStatus FireMaya::ToonMaterial::compute(const MPlug& plug, MDataBlock& block)
+{
+	if ((plug == Attribute::output) || (plug.parent() == Attribute::output))
```

Comment:
I'm not sure about this... Can you add a comment explaining the logic here?
E.g. in Uber we set Attribute::diffuseColor and outputAlpha here and in e.g. IBL we just return MS::kUnknownParameter...

Reply:
Here we just clean output attribute to make it "calculated", this is similar to PbrShader implementation.
Not a big deal regarding that actually. we don't make real compute here since all calculations occur during Render call only